api/src/main/java/jakarta/enterprise/invoke/InvokerBuilder.java

```diff
@@ -0,0 +1,363 @@
+ * locale, so probably not the best thing to do in real code, but enough for this
+ * example), and to transform the return value, we write a transformer as a simple
+ * {@code static} method:
+ *
+ * <pre>
+ * class Transformations {
+ *     static String repeatTwice(String str) {
+ *         return str + " " + str;
+ *     }
+ * }
+ * </pre>
+ *
+ * Then, assuming we have obtained the {@code InvokerBuilder} for {@code MyService.hello()},
+ * we can set up the lookup and transformations and build an invoker like so:
+ *
+ * <pre>
+ * builder.setInstanceLookup()
+ *        .setArgumentTransformer(0, String.class, "toUpperCase")
+ *        .setReturnValueTransformer(Transformations.class, "repeatTwice")
+ *        .build();
+ * </pre>
+ *
+ * The resulting invoker will be equivalent to the following class:
+ * <pre>
+ * class TheInvoker implements Invoker&lt;MyService, String&gt; {
+ *     String invoke(MyService ignored, Object[] arguments) {
+ *         MyService instance = CDI.current().select(MyService.class).get();
+ *         String argument = (String) arguments[0];
+ *         String transformedArgument = argument.toUpperCase();
+ *         String result = instance.hello(transformedArgument);
+ *         String transformedResult = Transformations.repeatTwice(result);
+ *         return transformedResult;
+ *     }
+ * }
+ * </pre>
+ *
+ * The caller of this invoker may pass {@code null} as the target instance, because
+ * the invoker will lookup the target instance on its own. Therefore, calling
+ * {@code invoker.invoke(null, new Object[] {"world"})} will return
+ * {@code "Hello WORLD! Hello WORLD!"}.
+ *
+ * <h2>General requirements</h2>
+ *
+ * To refer to a transformer or a wrapper, all methods in this builder accept:
+ * 1. the {@code Class} that that declares the method, and 2. the {@code String} name
+ * of the method.
+ * <p>
+ * Transformers may be {@code static}, in which case they must be declared directly
+ * on the given class, or they may be instance methods, in which case they may be declared
+ * on the given class or inherited from any of its supertypes.
+ * <p>
+ * It is possible to register only one transformer of each kind, or for each argument
+ * position in case of argument transformers. Attempting to register a second transformer
+ * of the same kind, or for the same argument position, leads to an exception.
+ * <p>
+ * Wrappers must be {@code static} and must be declared directly on the given class.
+ * It is possible to register only one wrapper. Attempting to register a second wrapper
+ * leads to an exception.
+ * <p>
+ * It is a deployment problem if no method with given name and valid signature is found,
+ * or if multiple methods with given name and different valid signatures are found. It is
+ * a deployment problem if a registered transformer or wrapper is not visible in the package
+ * that contains the class which declares the invokable method.
```

Comment:
This is not correct - the visibility of transformer/wrapper method from the package of the bean is irrelevant. Both are called independently by the impl which needs to look up those methods.
In fact, we should require transformers/wrappers to be `public`, unless we want the impl to enforce lookup via something like `MethodHandles.privateLookupIn(...)` which isn't very nice :)

Reply:
Yeah, in general, the author of the transformer and the author of the managed bean are different, so in practice, this rule collapses to "must be `public`" anyway. I agree we should just demand that transformers and wrappers are always `public`.